internal/provider/kubernetes/controller.go

```diff
@@ -1543,132 +1548,143 @@ func (r *gatewayAPIReconciler) enqueueClass(_ context.Context, _ client.Object)
 	}}}
 }
 
-func (r *gatewayAPIReconciler) hasManagedClass(ep *egv1a1.EnvoyProxy) bool {
-	// The EnvoyProxy must be in the same namespace as EG.
-	if ep.Namespace != r.namespace {
-		r.log.Info("envoyproxy namespace does not match Envoy Gateway's namespace",
-			"namespace", ep.Namespace, "name", ep.Name)
-		return false
+// processGatewayParamsRef processes the infrastructure.parametersRef of the provided Gateway.
+func (r *gatewayAPIReconciler) processGatewayParamsRef(ctx context.Context, gtw *gwapiv1.Gateway, resourceMap *resourceMappings, resourceTree *gatewayapi.Resources) error {
+	if gtw == nil || gtw.Spec.Infrastructure == nil || gtw.Spec.Infrastructure.ParametersRef == nil {
+		return nil
 	}
 
-	gcList := new(gwapiv1.GatewayClassList)
-	err := r.client.List(context.TODO(), gcList)
-	if err != nil {
-		r.log.Error(err, "failed to list gatewayclasses")
-		return false
+	if resourceTree.ClassEnvoyProxy != nil && resourceTree.ClassEnvoyProxy.Spec.MergeGateways != nil && *resourceTree.ClassEnvoyProxy.Spec.MergeGateways {
+		return fmt.Errorf("infrastructure.parametersref must be nil when MergeGateways feature is enabled by gatewayclass")
 	}
 
-	for i := range gcList.Items {
-		gc := gcList.Items[i]
-		// Reconcile the managed GatewayClass if it's referenced by the EnvoyProxy.
-		if r.hasMatchingController(&gc) &&
-			classAccepted(&gc) &&
-			classRefsEnvoyProxy(&gc, ep) {
-			return true
+	ref := gtw.Spec.Infrastructure.ParametersRef
+	if !(string(ref.Group) == egv1a1.GroupVersion.Group &&
```

Comment:
is there anyway to wrap these up into a helper function ? just like what we did in https://github.com/envoyproxy/gateway/blob/85cac8486dd8ba88841c36cd0cc65157c6bec7a2/internal/provider/kubernetes/helpers.go#L190